File: Automator/src/main/java/com/winkar/AppiumAgent.java
package com.winkar;

import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.android.AndroidElement;
import org.openqa.selenium.By;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.remote.DesiredCapabilities;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;

/**
 * Created by WinKaR on 16/3/14.
 */
public class AppiumAgent {
    private AndroidDriver driver;
    private int screenShotCounter;


    public AppiumAgent(String appPath) {
        try {
            DesiredCapabilities capabilities = new DesiredCapabilities();

            capabilities.setCapability("deviceName", "Galaxy Note4");
            capabilities.setCapability("platformVersion", "4.4");
            capabilities.setCapability("platformName", "Android");
            capabilities.setCapability("app", appPath);
            driver = new AndroidDriver(new URL("http://localhost:4723/wd/hub"), capabilities);

        } catch (MalformedURLException e) {
        }
    }

    public String formatAndroidElement(AndroidElement element) {
        StringBuilder sb = new StringBuilder();
        sb.append(element.getTagName())
                .append("\n")
                .append("\tId:" + element.getId())
                .append("\n")
                .append("\tText:" + element.getText())
                .append("\n\tClickable:" + element.getAttribute("clickable"))
                .append("\n");

        return sb.toString();
    }

    public void takeScreenShot(String logDir) {
        OutputStream os = null;
        try {
            byte[] screenShot = driver.getScreenshotAs(OutputType.BYTES);

            os = new FileOutputStream(
                    String.format("%s/%d.png", logDir, screenShotCounter)
            );
//            os = new FileOutputStream(logDir + File.separator + "screenshot" + Integer.toString(screenShotCounter) + ".png");
            screenShotCounter += 1;

            for (int i = 0; i < screenShot.length; i++) {
                os.write(screenShot[i]);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (os != null) {
                try {
                    os.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

    }

    public String currentActivity() {
        return driver.currentActivity();
    }

    public void pressKeyCode(int key) {
        driver.pressKeyCode(key);
    }

    public List findElements(By by) {
        return driver.findElements(by);
    }

    public void quit() {
        driver.quit();
    }

    public void installApp(String apkPath) {
        driver.installApp(apkPath);
    }

    public void removeApp(String bundleId) {
        driver.removeApp(bundleId);
    }

    public void startActivity(String appPackage, String activity) {
        driver.startActivity(appPackage, activity);
    }
}

File: Automator/src/main/scala/com/winkar/AppiumAgent.scala
package com.winkar

import java.net.URL
import java.nio.file.Paths

import io.appium.java_client.android.{AndroidDriver, AndroidElement}
import org.apache.commons.io.FileUtils
import org.apache.log4j.Logger
import org.openqa.selenium.remote.{CapabilityType, DesiredCapabilities}
import org.openqa.selenium.{By, OutputType}

import scala.collection.JavaConverters._


class AppiumAgent(val appPath: String) {
  val capabilities: DesiredCapabilities = new DesiredCapabilities
  capabilities.setCapability("deviceName", "Galaxy Note4")
  capabilities.setCapability("platformVersion", "4.4")
  capabilities.setCapability("platformName", "Android")
  capabilities.setCapability("app", appPath)
  val driver = new AndroidDriver[AndroidElement](new URL("http://localhost:4723/wd/hub"), capabilities)


  val log: Logger = Logger.getLogger(Automator.getClass.getName)
  var screenShotCounter: Int = 0



  def takeScreenShot(logDir: String) {
    val screenShotFile: Array[Byte] = driver.getScreenshotAs(OutputType.BYTES)

    FileUtils.writeByteArrayToFile(Paths.get(logDir, s"$screenShotCounter.png").toFile, screenShotFile)
    screenShotCounter += 1
  }

  def currentActivity: String = driver.currentActivity

  def currentPackage: String = AndroidUtils.getCurrentPackage

  def pressKeyCode(key: Int) = driver.pressKeyCode(key)


  def findElements(by: By): List[AndroidElement] = driver.findElements(by).asScala.toList

  def quit = driver.quit

  def closeApp = driver.closeApp

  def installApp(apkPath: String) = driver.installApp(apkPath)

  def removeApp(bundleId: String) = driver.removeApp(bundleId)

  def startActivity(appPackage: String, activity: String) = driver.startActivity(appPackage, activity)
}
